source/Coop.Core/Server/Services/Sync/Handlers/PeerQueueOverloadedHandler.cs

```diff
@@ -0,0 +1,114 @@
+﻿using Common;
+using Common.Messaging;
+using Common.Network;
+using Coop.Core.Server.Services.Time.Handlers;
+using Coop.Core.Server.Services.Time.Messages;
+using GameInterface.Services.GameDebug.Messages;
+using GameInterface.Services.Heroes.Enum;
+using LiteNetLib;
+using System;
+using System.Collections.Generic;
+namespace Coop.Core.Server.Services.Sync.Handlers
+{
+    internal class PeerQueueOverloadedHandler : IHandler
+    {
+        private static readonly long POLL_INTERVAL = 100;
+
+        private readonly IMessageBroker messageBroker;
+        private readonly INetwork network;
+
+        private readonly Poller poller;
+        private readonly HashSet<NetPeer> overloadedPeers;
+
+        private readonly TimeHandler timeHandler;
+
+        private TimeControlEnum originalSpeed;
+
+        public PeerQueueOverloadedHandler(
+            IMessageBroker messageBroker, 
+            INetwork network,
+            TimeHandler timeHandler
+        )
+        {
+            this.messageBroker = messageBroker;
+            this.network = network;
+            this.timeHandler = timeHandler;
+
+            overloadedPeers = new();
+            poller = new(Poll, TimeSpan.FromMilliseconds(POLL_INTERVAL));
+
+            messageBroker.Subscribe<PeerQueueOverloaded>(Handle);
+
+            poller.Start();
+        }
+
+        private void Handle(MessagePayload<PeerQueueOverloaded> payload)
+        {
+            lock (overloadedPeers)
+            {
+                if (overloadedPeers.Add(payload.What.NetPeer))
+                {
+                    if(timeHandler.TryGetTimeControlMode(out TimeControlEnum t))
+                    {
+                        originalSpeed = t;
+                    } else
+                    {
+                        originalSpeed = TimeControlEnum.Play_1x;
+                    }
+
+                    timeHandler.SetTimeMode(TimeControlEnum.Pause);
+
+                    var msg = new SendInformationMessage($"{overloadedPeers} clients are catching up, pausing");
+                    messageBroker.Publish(this, msg);
+                    network.SendAll(msg);
+
+                    poller.Start();
+                }
+            }
+        }
+
+        private void Poll(TimeSpan _)
+        {
+            List<NetPeer> toRemove = new();
+
+            foreach (var peer in new HashSet<NetPeer>(overloadedPeers))
+            {
+                if (peer.GetPacketsCountInReliableQueue(0, false) == 0)
+                {
+                    toRemove.Add(peer);
+                }
+            }
+
+            if (toRemove.Count == 0)
+            {
+                return;
+            }
```

Comment:
If you move this into the lock block, you do not need to create a copy of overloadedPeers `HashSet<NetPeer>(overloadedPeers)`

Reply:
I don't want to block the resource until I'm sure that there's something to change. I may reconsider.